### DISCUSSION TITLE
"refresh" button function?

### DISCUSSION THREAD

**Discussion by davecotter** [sqlitebrowser/sqlitebrowser#2648] [Q&A]
I thought the refresh button would reload from disk but .. am i wrong about that?
is there a "reload" function so i don't have to close the DB then open it again?

**Comment by chrisjlocke** [comment]
It should reload the data and display it when you view via the 'browse data' tab. If you use the 'execute sql' tab, the database is read each time.
What operating system are you using? What version of DB4S? 
So you're opening a database, viewing records, an external application is adding additional records, but DB4S isn't showing those additional records? Are you sure the external application is committing the records? Do you know what type of journal you're using?
If I open two copies of DB4S and open the same database, it works as desired...

**Reply by davecotter** [reply]
i'm on macOS (10.14.6), with app version 3.12.1. using regular journaling (not WAL mode).

1. i load a DB read-only, and yes i'm browsing in the "browse data" tab
2. i use my other app to change the data (not necessarily add / remove records, but modify existing ones), quit my app (so i KNOW the data is committed), then switch to DBBrowser, i click refresh, and it does NOT reload from disk. 
3. i close the db and open it again: NOW i see updated data

not realizing that it operated this way i spent a lot of time attempting to debug my app, only to realize later my app was working fine, and the refresh button was not doing what i expected.

**Reply by davecotter** [reply]
i'd REALLY expect that if i open a DB read only that it would auto-update, ie: watch the file for datestamp changes, and auto-reload if the date changes

**Reply by justinclift** [reply]
> i'd REALLY expect that if i open a DB read only that it would auto-update, ie: watch the file for datestamp changes, and auto-reload if the date changes

That's not a bad idea. :smile:

**Reply by chrisjlocke** [reply]
Similar to #2508 then....

**Reply by davecotter** [reply]
so, from what i gather, pressing the "reload" button should have the EXACT same effect as closing the database and re-opening it? cuz that's not what i'm seeing. i press reload and nothing happens. i close and reopen the database and NOW i see different data.

**Reply by chrisjlocke** [reply]
As per one of my other comments.  If you create one database and fire up two instances of DB4S, can you get them to see the same data (as per my screencast).  I'm trying to work out if its a MacOS thing, a database thing, or just something wacky on your particular installation.  It'll have to be a proper file database and not an in-memory database.  My screencast shows adding some records and this is reflected in the other instance.

**Reply by davecotter** [reply]
i'm not disputing your testimony, i'm just giving mine. as per one of my other comments, it's a disk database, not in-memory. the transaction is committed. i click refresh in the db: no refresh. i close and re-open: there IS a refresh. that's it. is there something more i can provide that will help this situation?

**Reply by chrisjlocke** [reply]
Oh no - I'm not disputing you at all either - just starting with what 'should' work, so we can work backwards to find where the problem lies - maybe its the database, a setting in it, or its DB4S, or its just a freaky installation.  If you can confirm you get a 'baseline' working as per my screencast, then we can try further on the real database, etc.  Just trying to get a baseline - not saying you're not getting an error, etc.  😊
Clicking refresh should not only redisplay the data, but rejiggle the schema, so the dropdown should show new tables, etc.  I'm on Windows, so know how it should work there, and we've not had other issues related to this, so I don't think its OS related.  I know we've had Linux issues like this, or where someone has connected to a remote Android database - the connection is then a bit different - caches in place, etc.
But this does sound weird...

**Comment by chrisjlocke** [comment]
DB4S doesn't 'auto update' (although that feature has been requested).  Just because the last modified date of the database has changed, doesn't mean the table the user is watching has changed.  A database might contain fifty tables.

For the avoidance of doubt, the refresh button does work as intended.  I even opened a database in read-only mode to test it wasn't an 'edge case' and that feature didn't work/was broken for read-only databases.

https://user-images.githubusercontent.com/7872218/112770269-d06a2280-901d-11eb-88a6-059add2d6abf.mp4

Are the databases local?  You say 'macOs' but it's not using some wacky file system is it?  I know there are issues with DB4S accessing databases across virtual file systems and the like, or on a remote device (eg, a phone).

**Comment by davecotter** [comment]
sure but instead of periodically refreshing, instead, refresh when the datestamp changes
modern file systems let you set up a listener on a file, so the OS notifies you when the file changes

**Reply by chrisjlocke** [reply]
But the filechange might not relate to that table.  If I'm adding a log record every second, but monitoring the doodah table, I don't want a refresh on the doodah table every second...

**Comment by ilovetogetspamed** [comment]
I know this is an old discussion but I use a Mac and had the same trouble with the 'refresh button'.

I had an instance of DB Browser for SQLite Version 3.12.2 (Built for x86_64-little_endian-lp64, running on x86_64, Qt Version 5.15.2, 
SQLCipher Version 4.4.3 community (based on SQLite 3.34.1)) running in a window and had the same DB open using Python Jupyter Notebook with PeeWee Orm to prototype a database I'm working on.

```python
...

db_filename = 'test.sqlite3'

# Connection that doesn't 'refresh' 
# db = SqliteDatabase(db_filename)

# Sqlite database using WAL-mode and 32MB page-cache.
db = SqliteDatabase(db_filename, pragmas={
    'journal_mode': 'wal',
    'cache_size': -32 * 1000})

...
```

The first connection method, i.e. without enabling WAL-mode, causes the refresh button to do nothing.

The second method works as expected.

I cant try it on Windoze, because I'm alergic to Windoze... ;-)

I hope this may help some poor soul having the same issue.

**Comment by bownie** [comment]
I noticed the same thing - first time user here on a Mac M1. I was expecting to see a refresh button on the first tab ("Database Structure") but only see it on the "Browse Data" tab. This does update table information when a table already exists but doesn't refresh database tables/structure. So, for example, if I create an empty DATABASE_DB file and open it in the browser, then create a table in the DB (via a separate process) and refresh in the "Browse Data" tab I don't see the table. I have to reopen the database.  

I'd love to see a "refresh structure" button on the first tab. If that's not planned currently would be happy to look into creating this as it would be really useful to me.

**Reply by MKleusberg** [reply]
Good point! I have just added such a button. Depending on your operating system you can either download a nightly build for Windows or macOS [here](https://nightlies.sqlitebrowser.org/latest/) by tomorrow or download an AppImage for Linux [here](https://github.com/sqlitebrowser/sqlitebrowser/releases/tag/continuous) in a couple of minutes.

**Comment by justinclift** [comment]
Sounds like the kind of thing that shouldn't be too hard for us to add a button for. :smile:

**Reply by muellert** [reply]
I'm using qlitebrowser to open a SQLite3 database read-only to inspect the data written by some apps during development. No amount of pressing "reload", F5 or executing SQL statements actually reloaded the database, so when my app eg. created additional tables, I could not see them unless I re-opened the database. Using read-only mode to avoid acquiring a lock on the database which would prevent the app from writing to it. So, having a button to reload the database structure and maybe also a button to toggle read-write, to cope with the locking issue, would be really beneficial.

I'm using this version of the software:

DB Browser for SQLite Version 3.12.2

Built for x86_64-little_endian-lp64, running on x86_64
Qt Version 5.15.4
SQLite Version 3.39.2.
